src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -456,9 +456,10 @@ private function validateAndDenormalize(array $types, string $currentClass, stri
                 // In XML and CSV all basic datatypes are represented as strings, it is e.g. not possible to determine,
                 // if a value is meant to be a string, float, int or a boolean value from the serialized representation.
                 // That's why we have to transform the values, if one of these non-string basic datatypes is expected.
+                $builtinType = $type->getBuiltinType();
                 if (\is_string($data) && (XmlEncoder::FORMAT === $format || CsvEncoder::FORMAT === $format)) {
                     if ('' === $data) {
-                        if (Type::BUILTIN_TYPE_ARRAY === $builtinType = $type->getBuiltinType()) {
+                        if (Type::BUILTIN_TYPE_ARRAY === $builtinType) {
                             return [];
                         }
 
```

Comment:
What about only replacing
```php
if ($type->isNullable() && \in_array($builtinType, [Type::BUILTIN_TYPE_BOOL, Type::BUILTIN_TYPE_INT, Type::BUILTIN_TYPE_FLOAT], true)) {
    return null;
}
```
by
```php
if ($type->isNullable() && Type::BUILTIN_TYPE_STRING === $builtinType) {
    return '';
}

if ($type->isNullable()) {
    return null;
}
```
instead?

This will simplify a lot the patch.

Reply:
That's not possible, because it needs to `return null` only when other options fail (like an object that accepts empty string as denormalization data). I've added `testDenormalizeUntypedStringObject` to test against this.